### PR TITLE
Feature: State Machine

### PR DESCRIPTION
**Pull request by JumpLink**
The main goal of this PR is to implement a `StateMachineService` as described in #25
* based on [valtio](https://github.com/pmndrs/valtio) (vanillla-version)
* Used to create a proxy state object using ` StateMachineService.create` for each chat session (sessions are handled using [Grammy's session plugin](https://grammy.dev/plugins/session))
* Changes on the state (or any sub property) can be subscribed using `StateMachineService.subscribe` or by watching the forwarded event

Other improvements: 
* Commands, Services and EventManagers now have a `getSingleton` (which internal resolves the DI singleton with the tools provided)
* In order to be able to cancel a running process such as answering callouts, we now make use of the native `AbortController`, this is assigned to a session so that it can be cancelled from anywhere
* Commands now have a new property `visibleOnStates` to define in which states this command should be visible
* New commands: `/help`, `/debug` and `/cancel`
* Renamed `TelegramService` to `AppService`
* Improved documentation
* Cleanups where it just fits

### CODE REVIEW

File: telegram-bot/services/communication.service.ts [hunk 1]
```diff
@@ -39,37 +47,40 @@ export class CommunicationService {
    * @param ctx
    * @param res
    */
-  public async send(ctx: Context, render: Render) {
+  public async send(ctx: AppContext, render: Render) {
+    const markup = render.keyboard ||
+      (render.removeKeyboard ? { remove_keyboard: true as true } : undefined);
```
Comment (wpf500) [inline telegram-bot/services/communication.service.ts:52]: Why `true as true`?

File: telegram-bot/types/chat-state.ts
Comment (wpf500) [inline telegram-bot/types/chat-state.ts:1]: Not sure what this file is for? Can't you just use the enum?

**Review by wpf500 — APPROVED** [review]
I obviously don't know the codebase well but from what I can tell it all looks great :smile:

File: telegram-bot/services/communication.service.ts (hunk 1 above)
Comment (JumpLink) [reply telegram-bot/services/communication.service.ts:52]: The type only allows `true` and not `boolean` but `true` is handled as `boolean` otherwise

**Pull request merged**